addons/test_mail/tests/test_message_post.py

```diff
@@ -41,6 +42,32 @@ def test_message_post_return(self):
         messageId = call_kw(self.env['mail.channel'], 'message_post', [test_channel.id], {'body': 'test'})
         self.assertTrue(isinstance(messageId, int))
 
+    @users('employee')
+    def test_notify_prepare_template_context_company_value(self):
+        """Verify that the template context company value is right
+        after switching the env company or if a company_id is set
+        on mail record."""
+        user = self.env.user
+        partner = user.partner_id
+        mainCompany = user.company_id
+        otherCompany = self.env['res.company'].with_user(self.user_admin).create({'name': 'Company B'})
+        user.sudo().write({'company_ids': [(4, otherCompany.id)]})
+
+        # self.env.company.id = Main Company    AND    partner.company_id = False
+        self.assertEqual(self.env.company.id, mainCompany.id)
+        template_values = partner._notify_prepare_template_context(partner.message_ids, {})
+        self.assertEqual(template_values.get('company').id, self.env.company.id)
+
+        # self.env.company.id = Other Company    AND    partner.company_id = False
```

Comment:
Could be great to assert that "test_record.company_id" is False (valid for all other comments: as this is a corner case fix, let us ensure our test hypothesis are correct beforehand